protocol/binary_test.go

```diff
@@ -788,3 +821,219 @@ func TestBinaryEnvelopeSuccessful(t *testing.T) {
 	}
 }
+
+func TestReqRes(t *testing.T) {
+
+	tests := []struct {
+		msg           string
+		req           wire.Value
+		reqBytes      []byte
+		responderType reflect.Type
+		res           wire.Value
+		resType       wire.EnvelopeType
+		resBytes      []byte
+	}{
+		{
+			msg:           "empty req, empty reply, no envelope",
+			req:           vstruct(),
+			reqBytes:      []byte{0x00},
+			responderType: reflect.TypeOf(noEnvelopeResponder),
+			res:           vstruct(),
+			resType:       wire.Reply,
+			resBytes:      []byte{0x00},
+		},
+		{
+			msg: "two field req, empty reply, no envelope",
+			req: vstruct(vfield(1, vbool(true))),
+			reqBytes: []byte{
+				0x02,       // type:1 = bool
+				0x00, 0x01, // id:2 = 1
+				0x01, // value = true
+				0x00, // stop
+			},
+			responderType: reflect.TypeOf(noEnvelopeResponder),
+			res:           vstruct(),
+			resType:       wire.Reply,
+			resBytes:      []byte{0x00},
+		},
+		{
+			msg: "empty reply, no-version non-strict envelope",
+			reqBytes: []byte{
+				0x00, 0x00, 0x00, 0x03, 'a', 'b', 'c', // name~4 = "abc"
+				0x01,                   // type:1 = 1 = call
+				0x00, 0x00, 0x15, 0x3c, // seqID:4 = 5436
+
+				// <struct>
+				0x06,       // type:1 = i16
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x64, // value = 100
+				0x00, // stop
+			},
+			req: vstruct(
+				vfield(1, vi16(100)),
+			),
+			responderType: reflect.TypeOf((*envelopeV0Responder)(nil)),
+			resType:       wire.Exception,
+			res:           vstruct(),
+			resBytes: []byte{
+				0x00, 0x00, 0x00, 0x03, 'a', 'b', 'c', // name~4 = "abc"
+				0x03,                   // type1 = 3 (exception)
+				0x00, 0x00, 0x15, 0x3c, // seqID:4 = 5436
+
+				// <struct>
+				0x00,
+			},
+		},
+		{
+			msg: "empty reply, version 1 strict envelope",
+			reqBytes: []byte{
+				0x80, 0x01, 0x00, 0x01, // version|type:4 = 1 | call
+				0x00, 0x00, 0x00, 0x03, 'a', 'b', 'c', // name~4 = "abc"
+				0x00, 0x00, 0x15, 0x3c, // seqID:4 = 5436
+
+				// <struct>
+				0x06,       // type:1 = i16
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x64, // value = 100
+				0x00, // stop
+			},
+			req: vstruct(
+				vfield(1, vi16(100)),
+			),
+			responderType: reflect.TypeOf((*envelopeV1Responder)(nil)),
+			resType:       wire.Reply,
+			res:           vstruct(),
+			resBytes: []byte{
+				0x80, 0x01, 0x00, 0x02, // version:2 &^ 0x80 = 1, type:2 = 2 (reply)
+				0x00, 0x00, 0x00, 0x03, 'a', 'b', 'c', // name~4 = "abc"
+				0x00, 0x00, 0x15, 0x3c, // seqID:4 = 5436
+
+				// <struct>
+				0x00,
+			},
+		},
+		{
+			msg: "complex request, no envelope, complex response",
+			req: vstruct(
+				vfield(1, vi16(42)),
+				vfield(2, vlist(wire.TBinary, vbinary("foo"), vbinary("bar"))),
+				vfield(3, vset(wire.TBinary, vbinary("baz"), vbinary("qux"))),
+			),
+			reqBytes: []byte{
+				0x06,       // type:1 = i16
+				0x00, 0x01, // id:2 = 1
+				0x00, 0x2a, // value = 42
+
+				0x0F,       // type:1 = list
```

Comment:
can we avoid manually encoding these bytes, and use the existing binary encoding? prefer to avoid re-testing the struct binary encoding here, which we already test elsewhere. the binary that is relevant here is the envelope, the rest should be `encode(vstruct(..))`